semanticdb/scalalib/gen/src/main/scala/scala/meta/internal/scalalib/Generator.scala

```diff
@@ -0,0 +1,169 @@
+      builtinMethod("AnyRef", List(p.FINAL), "ne", Nil, List("that" -> "_root_.scala.AnyRef#"), "_root_.scala.Boolean#"),
+      builtinMethod("AnyRef", List(p.FINAL), "synchronized", List("T"), List("body" -> "_root_.scala.AnyRef.synchronized(T).[T]"), "_root_.scala.AnyRef.synchronized(T).[T]"))
+    builtinClass(Nil, "AnyRef", List("_root_.scala.Any#"), symbols.flatten)
+  }
+
+  def nothing: ToplevelInfos = {
+    builtinClass(List(p.ABSTRACT, p.FINAL), "Nothing", List("_root_.scala.Any#"), Nil)
+  }
+
+  private def builtinClass(
+      props: List[s.SymbolInformation.Property],
+      name: String,
+      bases: List[String],
+      symbols: List[s.SymbolInformation]): ToplevelInfos = {
+    val parents = bases.map { base =>
+      s.Type(tag = t.TYPE_REF, typeRef = Some(s.TypeRef(None, base, Nil)))
+    }
+    val symbol = "_root_.scala." + name + "#"
+    val builtinTpe = s.Type(tag = t.TYPE_REF, typeRef = Some(s.TypeRef(None, symbol, Nil)))
+    val ctorSig = s.MethodType(Nil, List(s.MethodType.ParameterList(Nil)), Some(builtinTpe))
+    val ctor = s.SymbolInformation(
+      symbol = symbol + "`<init>`().",
+      language = Some(builtinLanguage),
+      kind = k.PRIMARY_CONSTRUCTOR,
+      name = "<init>",
+      tpe = Some(s.Type(tag = t.METHOD_TYPE, methodType = Some(ctorSig))),
+      accessibility = Some(s.Accessibility(a.PUBLIC)),
+      owner = symbol
+    )
+    val builtinSig = {
+      val decls = symbols.filter(_.kind == k.DEF)
+      val tpe = s.ClassInfoType(Nil, parents, ctor.symbol +: decls.map(_.symbol))
+      s.Type(tag = t.CLASS_INFO_TYPE, classInfoType = Some(tpe))
+    }
+    val builtin = s.SymbolInformation(
+      symbol = symbol,
+      language = Some(builtinLanguage),
+      kind = k.CLASS,
+      properties = props.foldLeft(0)((acc, prop) => acc | prop.value),
+      name = name,
+      tpe = Some(builtinSig),
+      accessibility = Some(s.Accessibility(a.PUBLIC)),
+      owner = "_root_.scala."
+    )
+    val syntheticBase = Paths.get(".")
+    val syntheticPath = syntheticBase.resolve("scala/" + name + ".class")
+    val syntheticClassfile = ToplevelClassfile(syntheticBase, syntheticPath, null)
+    ToplevelInfos(syntheticClassfile, List(builtin), ctor +: symbols)
+  }
+
+  def builtinMethod(
+      className: String,
+      props: List[s.SymbolInformation.Property],
+      methodName: String,
+      tparamDsls: List[String],
+      paramDsls: List[(String, String)],
+      retTpeSymbol: String): List[s.SymbolInformation] = {
+    val classSymbol = "_root_.scala." + className + "#"
+    val encodedMethodName = {
+      if (Character.isJavaIdentifierStart(methodName.head)) methodName
+      else "`" + methodName + "`"
+    }
+    val disambiguator = {
+      val paramTypeDescriptors = paramDsls.map(_._2).map { symbol =>
+        // TODO: It would be nice to have a symbol parser in semanticdb3.
```

Comment:
I'm concerned about the spread of logic around handling symbol strings. I'm not convinced this bare bones approach is sustainable given how easy it is to mess up and lack of static checks to help you prevent do mistakes.

Reply:
Yep. We'll need to refactor all that once the dust settles. I'm not doing the refactoring right now to be sure we collect enough use cases and don't overengineer the API.

Reply:
https://github.com/scalameta/scalameta/issues/1381